### PR TITLE
Fix memcpy sizes in gpujpeg_table_huffman_*_init()

### PR DESCRIPTION
**Pull request by xqms**
I got the following compiler warnings in the current master branch:

```
In file included from /usr/include/string.h:495,
                 from /home/user/schwarzm/contrib/GPUJPEG/src/gpujpeg_util.h:37,
                 from /home/user/schwarzm/contrib/GPUJPEG/src/gpujpeg_table.c:32:
In function ‘memcpy’,
    inlined from ‘gpujpeg_table_huffman_encoder_init’ at /home/user/schwarzm/contrib/GPUJPEG/src/gpujpeg_table.c:314:13:
/usr/include/x86_64-linux-gnu/bits/string_fortified.h:34:10: warning: ‘__builtin_memcpy’ reading 256 bytes from a region of size 12 [-Wstringop-overflow=]
   34 |   return __builtin___memcpy_chk (__dest, __src, __len, __bos0 (__dest));
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In function ‘memcpy’,
    inlined from ‘gpujpeg_table_huffman_encoder_init’ at /home/user/schwarzm/contrib/GPUJPEG/src/gpujpeg_table.c:317:13:
/usr/include/x86_64-linux-gnu/bits/string_fortified.h:34:10: warning: ‘__builtin_memcpy’ reading 256 bytes from a region of size 162 [-Wstringop-overflow=]
   34 |   return __builtin___memcpy_chk (__dest, __src, __len, __bos0 (__dest));
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In function ‘memcpy’,
    inlined from ‘gpujpeg_table_huffman_encoder_init’ at /home/user/schwarzm/contrib/GPUJPEG/src/gpujpeg_table.c:322:13:
/usr/include/x86_64-linux-gnu/bits/string_fortified.h:34:10: warning: ‘__builtin_memcpy’ reading 256 bytes from a region of size 12 [-Wstringop-overflow=]
   34 |   return __builtin___memcpy_chk (__dest, __src, __len, __bos0 (__dest));
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In function ‘memcpy’,
    inlined from ‘gpujpeg_table_huffman_encoder_init’ at /home/user/schwarzm/contrib/GPUJPEG/src/gpujpeg_table.c:325:13:
/usr/include/x86_64-linux-gnu/bits/string_fortified.h:34:10: warning: ‘__builtin_memcpy’ reading 256 bytes from a region of size 162 [-Wstringop-overflow=]
   34 |   return __builtin___memcpy_chk (__dest, __src, __len, __bos0 (__dest));
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In function ‘memcpy’,
    inlined from ‘gpujpeg_table_huffman_decoder_init’ at /home/user/schwarzm/contrib/GPUJPEG/src/gpujpeg_table.c:342:13:
/usr/include/x86_64-linux-gnu/bits/string_fortified.h:34:10: warning: ‘__builtin_memcpy’ reading 256 bytes from a region of size 12 [-Wstringop-overflow=]
   34 |   return __builtin___memcpy_chk (__dest, __src, __len, __bos0 (__dest));
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In function ‘memcpy’,
    inlined from ‘gpujpeg_table_huffman_decoder_init’ at /home/user/schwarzm/contrib/GPUJPEG/src/gpujpeg_table.c:345:13:
/usr/include/x86_64-linux-gnu/bits/string_fortified.h:34:10: warning: ‘__builtin_memcpy’ reading 256 bytes from a region of size 162 [-Wstringop-overflow=]
   34 |   return __builtin___memcpy_chk (__dest, __src, __len, __bos0 (__dest));
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In function ‘memcpy’,
    inlined from ‘gpujpeg_table_huffman_decoder_init’ at /home/user/schwarzm/contrib/GPUJPEG/src/gpujpeg_table.c:350:13:
/usr/include/x86_64-linux-gnu/bits/string_fortified.h:34:10: warning: ‘__builtin_memcpy’ reading 256 bytes from a region of size 12 [-Wstringop-overflow=]
   34 |   return __builtin___memcpy_chk (__dest, __src, __len, __bos0 (__dest));
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In function ‘memcpy’,
    inlined from ‘gpujpeg_table_huffman_decoder_init’ at /home/user/schwarzm/contrib/GPUJPEG/src/gpujpeg_table.c:353:13:
/usr/include/x86_64-linux-gnu/bits/string_fortified.h:34:10: warning: ‘__builtin_memcpy’ reading 256 bytes from a region of size 162 [-Wstringop-overflow=]
   34 |   return __builtin___memcpy_chk (__dest, __src, __len, __bos0 (__dest));
      |
```

Fix that by using the smaller source buffer size.

Also adds asserts to check that the buffers are large enough.

### CODE REVIEW

**Comment by xqms** [comment]
By the way, I forgot to say: Awesome project! Using this I was able to shave off ~10ms in comparison to my previous streaming solution using nvJPEG :+1:

File: src/gpujpeg_table.c [hunk 1]
```diff
@@ -308,21 +308,30 @@ gpujpeg_table_huffman_encoder_init(struct gpujpeg_table_huffman_encoder* table,
 {
     assert(comp_type == GPUJPEG_COMPONENT_LUMINANCE || comp_type == GPUJPEG_COMPONENT_CHROMINANCE);
     assert(huff_type == GPUJPEG_HUFFMAN_DC || huff_type == GPUJPEG_HUFFMAN_AC);
+    assert(sizeof(table->bits) >= gpujpeg_table_huffman_y_dc_bits);
```
Comment (MartinPulec) [inline src/gpujpeg_table.c:311]: there is missing sizeof around second argument

File: src/gpujpeg_table.c [hunk 2]
```diff
@@ -308,21 +308,30 @@ gpujpeg_table_huffman_encoder_init(struct gpujpeg_table_huffman_encoder* table,
...
+    assert(sizeof(table->huffval) >= sizeof(gpujpeg_table_huffman_y_dc_value));
+    assert(sizeof(table->bits) >= sizeof(gpujpeg_table_huffman_y_ac_bits));
+    assert(sizeof(table->huffval) >= sizeof(gpujpeg_table_huffman_y_ac_value));
+    assert(sizeof(table->bits) >= gpujpeg_table_huffman_cbcr_dc_bits);
```
Comment (MartinPulec) [inline src/gpujpeg_table.c:315]: ditto

File: src/gpujpeg_table.c [hunk 3]
```diff
@@ -336,21 +345,30 @@ gpujpeg_table_huffman_decoder_init(struct gpujpeg_table_huffman_decoder* table,
 {
     assert(comp_type == GPUJPEG_COMPONENT_LUMINANCE || comp_type == GPUJPEG_COMPONENT_CHROMINANCE);
     assert(huff_type == GPUJPEG_HUFFMAN_DC || huff_type == GPUJPEG_HUFFMAN_AC);
+    assert(sizeof(table->bits) >= gpujpeg_table_huffman_y_dc_bits);
```
Comment (MartinPulec) [inline src/gpujpeg_table.c:348]: "

File: src/gpujpeg_table.c [hunk 4]
```diff
@@ -336,21 +345,30 @@ gpujpeg_table_huffman_decoder_init(struct gpujpeg_table_huffman_decoder* table,
...
+    assert(sizeof(table->huffval) >= sizeof(gpujpeg_table_huffman_y_dc_value));
+    assert(sizeof(table->bits) >= sizeof(gpujpeg_table_huffman_y_ac_bits));
+    assert(sizeof(table->huffval) >= sizeof(gpujpeg_table_huffman_y_ac_value));
+    assert(sizeof(table->bits) >= gpujpeg_table_huffman_cbcr_dc_bits);
```
Comment (MartinPulec) [inline src/gpujpeg_table.c:352]: "

**Comment by MartinPulec** [comment]
Hi, thanks, looks quite fine - I think that there are missing sizeof with some arguments (see coments in code; won't those asserts always fail?). Also it would be nice if there were _Static_asserts instead of plain assert (although I bet that even normal assert will be compiled out). Do you want to look at it by yourself or should I after the merge?

File: src/gpujpeg_table.c (hunk 1 above)
Comment (xqms) [reply src/gpujpeg_table.c]: That's what I get for adding asserts in the last minute ;) Fixed.

**Comment by xqms** [comment]
I fixed the missing sizeofs() (it was still a valid expression, even if it didn't make sense at all). I'm a C++ guy, where we have static_assert - I wasn't aware that C has `_Static_assert` as well. Everyday you learn something new ;)

**Comment by MartinPulec** [comment]
> I'm a C++ guy, where we have static_assert - I wasn't aware that C has _Static_assert as well.

Well, there is even `static_assert` in current C directly, but it is a macro defined in assert.h for the above keyword.

**Pull request merged**